### PR TITLE
Don’t duplicate datasets with a `harvest.remote_id` URI when harvesting

### PR DESCRIPTION
**Pull request by magopian**
Fixes [#1587](https://github.com/datagouv/data.gouv.fr/issues/1587)

### CODE REVIEW

File: docs/harvesting.md [hunk 1]
```diff
@@ -89,6 +89,12 @@ optional arguments:
 ## Backends
 
 `udata` comes with 3 harvest backends but you can implement your own backend.
```
Comment (bolinocroustibat) [inline docs/harvesting.md:91]: Would be nice to list the existing backends in the README here, no?

**Review by bolinocroustibat — APPROVED** [review]
LGTM, looks like it makes the code more robust. Always painful to implement a test, except maybe when it is to test a very new feature :)

File: docs/harvesting.md [hunk 2]
```diff
@@ -251,6 +257,22 @@ This will create a new package with a harvester skeleton on which you can start
 You may take a look at the [existing backends][backends-repository] to see exiting implementations.
 
 
+## Debugging
```
Comment (maudetes) [inline docs/harvesting.md:260]: I think you may want to document this dedicated command instead?
```
udata harvest run <source_id>
```

File: udata/harvest/tests/test_base_backend.py [hunk 3]
```diff
@@ -281,6 +287,78 @@ def test_harvest_datasets_get_deleted(self):
         for item in job.items:
             assert item.dataset is None
 
+    def test_no_datasets_duplication(self, app):
+        limit = app.config["HARVEST_AUTOARCHIVE_GRACE_DAYS"]
+        last_update = datetime.utcnow() - timedelta(days=limit - 1)
+        reuse_remote_id_uri = "http://example.com/reuse-this-uri"
```
Comment (maudetes) [inline udata/harvest/tests/test_base_backend.py:293]: `reuse` means something else in the udata context, it may be misinterpreted :stuck_out_tongue: 
`duplicated_remote_id_uri` maybe?

File: udata/harvest/tests/test_base_backend.py [hunk 4]
```diff
@@ -281,6 +287,78 @@ def test_harvest_datasets_get_deleted(self):
...
+        nb_datasets = 3
+        source = HarvestSourceFactory(
+            config={
+                "nb_datasets": nb_datasets,
+                "dataset_remote_ids": [reuse_remote_id_uri],
```
Comment (maudetes) [inline udata/harvest/tests/test_base_backend.py:298]: Don't we want to have either `nb_datasets` OR `dataset_remote_ids`?
Else I find it hard to follow what happens when we have both :)

File: udata/harvest/tests/test_base_backend.py [hunk 5]
```diff
@@ -281,6 +287,78 @@ def test_harvest_datasets_get_deleted(self):
...
+            harvest={
+                "domain": source.domain,
+                "remote_id": "fake-0",  # the FakeBackend harvest should reuse this dataset
+                "source_id": str(source.id),
+                "last_update": last_update,
+            },
+        )
+
+        # Create a dataset that should be reused even though it's a different `remote_id` and `domain,
+        # because it's an URI.
```
Comment (maudetes) [inline udata/harvest/tests/test_base_backend.py:316]: Did you mean?
```suggestion
        # Create a dataset that should be reused even though it's a different `source_id` and `domain,
        # because the remote_id is the same and an URI.
```

File: udata/harvest/tests/test_base_backend.py [hunk 6]
```diff
@@ -281,6 +287,78 @@ def test_harvest_datasets_get_deleted(self):
         for item in job.items:
             assert item.dataset is None
 
+    def test_no_datasets_duplication(self, app):
+        limit = app.config["HARVEST_AUTOARCHIVE_GRACE_DAYS"]
+        last_update = datetime.utcnow() - timedelta(days=limit - 1)
```
Comment (maudetes) [inline udata/harvest/tests/test_base_backend.py:292]: I think bringing `HARVEST_AUTOARCHIVE_GRACE_DAYS` here makes it seem like it has some impact, but it should not right?

**Review by maudetes — APPROVED** [review]
Thank you!
And thank you for adding tests, it's not always easy with harvesting.
I have minor comments only

Don't forget the changelog ;)

File: docs/harvesting.md (hunk 1 above)
Comment (magopian) [reply docs/harvesting.md:91]: yes indeed, i'll add the ones i'm aware of ;)

File: docs/harvesting.md (hunk 2 above)
Comment (magopian) [reply docs/harvesting.md:260]: Incredible, this was there all along, I didn't realize 😬

File: udata/harvest/tests/test_base_backend.py (hunk 3 above)
Comment (magopian) [reply udata/harvest/tests/test_base_backend.py:293]: yes indeed, confusing, i'll change it, thanks!

File: udata/harvest/tests/test_base_backend.py (hunk 5 above)
Comment (magopian) [reply udata/harvest/tests/test_base_backend.py:316]: Yeah, it's wayyyyy clearer that way, thanks 😅

File: udata/harvest/tests/test_base_backend.py (hunk 6 above)
Comment (magopian) [reply udata/harvest/tests/test_base_backend.py:292]: Totally right! It was some leftover from a previous copy-paste, well spotted, thanks, i'll remove it straight away 🙏

File: udata/harvest/tests/test_base_backend.py (hunk 4 above)
Comment (magopian) [reply udata/harvest/tests/test_base_backend.py]: I updated the tests: https://github.com/opendatateam/udata/pull/3219/commits/aec0a6a5bf36e5a866bb7b4b6c73f50db47786aa
What do you think?

Comment (maudetes) [reply udata/harvest/tests/test_base_backend.py]: Yes, I find it better, thank you!
